subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/RerunPreviousFailedTestIntegrationTest.groovy

```diff
@@ -0,0 +1,173 @@
+
+    def setup() {
+        buildFile << """
+            apply plugin: 'java'
+
+            ${mavenCentralRepository()}
+
+
+            dependencies {
+                testCompile 'junit:junit:4.12'
+            }
+
+            tasks.withType(Test) {
+                maxParallelForks = 1 
+            }
+        """
+
+        (1..3).each {
+            file("src/test/java/ConditionalFailingTest_${it}.java") << """
+                import org.junit.Test;
+                public class ConditionalFailingTest_${it} {
+                    @Test
+                    public void failedTest() {
+                        if("${it}".equals(System.getProperty("${INDEX_OF_TEST_TO_FAIL}"))) {
+                            throw new RuntimeException();
+                        }
+                    }
+                }
+            """.stripIndent()
+        }
+    }
+
+    def letTestFail(def index) {
+        buildFile << """
+        test {
+            systemProperty('${INDEX_OF_TEST_TO_FAIL}', '${index}')
+        }
+        """
+    }
+
+    @Unroll
+    def 'subsequent fail fast execution runs failed test first'() {
+        given:
+        letTestFail(indexOfTestToFail)
+
+        when:
+        fails('test')
+
+        then:
+        testFailed(indexOfTestToFail)
+
+        when:
+        fails('test', '--fail-fast')
+
+        then:
+        testFailedAndOthersIgnoredOrNotExecuted(indexOfTestToFail)
+
+        when:
+        letTestFail(0)
+        succeeds('test')
+
+        then:
+        allTestsSucceed()
+
+        where:
+        indexOfTestToFail << [1, 2, 3]
+    }
+
+    @Unroll
+    def 'can delete previous failed test'() {
+        given:
+        letTestFail(indexOfTestToFail)
+
+        when:
+        fails('test')
+
+        then:
+        testFailed(indexOfTestToFail)
+
+        when:
+        file("src/test/java/ConditionalFailingTest_${indexOfTestToFail}.java").delete()
+        succeeds('test')
+
+        then:
+        remainTestsSucceed(indexOfTestToFail)
+
+        where:
+        indexOfTestToFail << [1, 2, 3]
+    }
+
+    @Unroll
+    def 'can modify previous failed test'() {
+        given:
+        letTestFail(indexOfTestToFail)
+
+        when:
+        fails('test')
+
+        then:
+        testFailed(indexOfTestToFail)
+
+        when:
+        file("src/test/java/ConditionalFailingTest_${indexOfTestToFail}.java").text = """
+        public class ConditionalFailingTest_${indexOfTestToFail} {
+            public void failedTest() {
+            }
+        }
+        """
+        succeeds('test')
+
+        then:
+        remainTestsSucceed(indexOfTestToFail)
+
+        where:
+        indexOfTestToFail << [1, 2, 3]
+    }
+
+    void testFailed(def indexOfTestToFail) {
+        new DefaultTestExecutionResult(testDirectory)
+            .testClass("ConditionalFailingTest_${indexOfTestToFail}").assertTestFailed('failedTest', Matchers.anything())
+    }
+
+    void allTestsSucceed() {
```

Comment:
⭕️ There is a lot of coupling in these methods - If I add a test number "4" in the tests above, it won't actually be checked because this is hard-coded to check 1, 2 and 3. Should probably be a shared constant that all these methods use.

Reply:
Thanks for your suggestion.